src/library/scala/concurrent/impl/ExecutionContextImpl.scala

```diff
@@ -53,39 +44,54 @@ private[concurrent] object ExecutionContextImpl {
       thread
     }
 
-    // As per ThreadFactory contract newThread should return `null` if cannot create new thread.
-    def newThread(runnable: Runnable): Thread =
-      if (reserveThread())
-        wire(new Thread(new Runnable {
-          // We have to decrement the current thread count when the thread exits
-          override def run() = try runnable.run() finally deregisterThread()
-        })) else null
-
-    def newThread(fjp: ForkJoinPool): ForkJoinWorkerThread =
-      if (reserveThread()) {
-        wire(new ForkJoinWorkerThread(fjp) with BlockContext {
-          // We have to decrement the current thread count when the thread exits
-          final override def onTermination(exception: Throwable): Unit = deregisterThread()
-          final override def blockOn[T](thunk: =>T)(implicit permission: CanAwait): T = {
-            var result: T = null.asInstanceOf[T]
-            ForkJoinPool.managedBlock(new ForkJoinPool.ManagedBlocker {
-              @volatile var isdone = false
-              override def block(): Boolean = {
-                result = try {
-                    // When we block, switch out the BlockContext temporarily so that nested blocking does not created N new Threads
-                    BlockContext.withBlockContext(BlockContext.defaultBlockContext) { thunk }
-                  } finally {
-                    isdone = true
-                  }
-
-                true
+    private[this] final def shouldAllocateUnlimited(): Boolean =
+      Thread.currentThread match {
+        case w: WorkerThread if w.isBlocking => false
+        case _ => true
+      }
+
+    override def newThread(runnable: Runnable): Thread =
+      if (shouldAllocateUnlimited()) wire(new Thread(runnable))
```

Comment:
Should it be really unlimited? Can it cause uncontrolled number of threads?

Reply:
This is actually harmless in this case since this thread factory is only going to be used for newThread(ForkJoinPool). The ThreadFactory is legacy compliance (when the fallback was a ThreadPoolExecutorService)

Reply:
ok, there is historical knowledge, which I do not know about, so do not take my review as a serious check.